tools/code_analyzer/op_dependency.cpp

```diff
@@ -0,0 +1,396 @@
+        (*functionsToSchemaStrs)[parentFunctionName].insert(schemaStr);
+        continue;
+      }
+
+      // Check pattern to register ops.
+      if (checkInstructionCalledFunction(
+          *cur, "c10::RegisterOperators::Options::schema")) {
+        if (seenOptionsSchemaCall) break;
+        seenOptionsSchemaCall = true;
+        continue;
+      }
+
+      // Collect function references between two op registration (between two
+      // "Options::schema" calls).
+      if (seenOptionsSchemaCall) {
+        auto cb = [&](Function* func) -> void {
+          (*schemaStrToFunctions)[schemaStr].insert(func->getName());
+          if (Verbose) {
+            std::cerr << "[DEBUG] " << schemaStr << " => "
+                      << demangle(func->getName()) << std::endl;
+          }
+        };
+        if (Verbose) {
+          std::cerr << "[DEBUG] " << schemaStr << " [INST] "
+                    << *cur << std::endl;
+        }
+        scanReferredFunctionsFromOperands(*cur, cb);
+      }
+    }
+  }
+
+  // llvm::CallGraph only checks CallSites. This method also checks constant
+  // function pointer references for better recall - the function might not
+  // actually be called.
+  static void scanReferenceGraph(Module& M, GRAPH* functionToFunctions) {
+    for (Function& F : M) {
+      std::string name = F.getName();
+      scanReferredFunctionsFromOperandsInFunction(F,
+          [&](Function* func) -> void {
+            (*functionToFunctions)[name].insert(func->getName());
+      });
+    }
+  }
+
+  static void scanFunctionsForDebugFilters(Module& M, SET* keyNodes) {
+    for (Function& F : M) {
+      std::string name = F.getName();
+      if (matchDebugFilters(name)) {
+        keyNodes->insert(name);
+      }
+    }
+  }
+
+  // Calculate transitive closure and remove non-key nodes.
+  static void simplifyGraph(GRAPH& input, SET& keyNodes, GRAPH* output) {
+    // Starting from every key node, use BFS to traverse all nodes that are
+    // transitively reachable from the node in the sparse graph.
+    for (auto& key : keyNodes) {
+      std::deque<std::string> queue;
+      SET expanded;  // has some runtime issue with std::unordered_set
+      auto expand = [&](const std::string& curNode) -> void {
+        if (!expanded.insert(curNode).second) return;
+        for (auto& next : input[curNode]) {
+          queue.emplace_back(next);
+        }
+      };
+
+      expand(key);
+      while (!queue.empty()) {
+        auto curNode = queue.front();
+        queue.pop_front();
+        if (keyNodes.count(curNode)) {
+          // Output links between key nodes.
+          (*output)[key].insert(curNode);
+          // Stop expanding key nodes.
+          if (!TransitiveClosure) continue;
+        }
+        // Expand node.
+        expand(curNode);
+      }
+    }
+  }
+
+  static bool isFunctionSchemaString(const std::string& str) {
+    return str.find("aten::") == 0 || str.find("quantized::") == 0;
```

Comment:
I know y'all have said you don't care about false positives, but I am thinking about the poor future engineer who causes space size to regress with a seemingly innocuous change, and it turns out it's because they had the misfortune of adding "aten::" to the beginning of a string and then suddenly this part of the code was considered a root and pulled a whole pile of other stuff. I don't know if there is a way to do this analysis 100% soundly, or if it's worth doing so. But I am still thinking of this engineer...

Reply:
strings starting from "aten::" won't have any effect unless the enclosing function also calls "c10::findSchema()" and/or "c10::schema()"...
My current plan is to check in the generated dependency graph and update it periodically (nightly job?)
We probably can check the diffs generated by the job? The mobile team will commit to monitor and fix whenever the size of dependency file changes dramatically.

Reply:
Yeah, running the dep graph nightly with monitoring seems reasonable. But if you run it every commit (doesn't seem like it takes long to run) you'll get to avoid having to bisect when something funny happens :)